app/src/main/java/com/android/calendar/Utils.java

```diff
@@ -793,8 +793,9 @@ public static long convertAlldayUtcToLocal(Time recycle, long utcTime, String tz
         }
         recycle.setTimezone(Time.TIMEZONE_UTC);
         recycle.set(utcTime);
-        recycle.setTimezone(tz);
-        return recycle.normalize();
+        Time target = new Time(tz);
```

Comment:
`recycle` (the time object "passed to the the function and if not created") contains the date of the all day event (with hour, minute and seconds set to zero, in other words time 00:00:00) in the UTC timezone.

`target` (the new time object created here) is set to the same date (still at time 00:00:00) but in the timezone `tz`.

The old code does:

```
recycle.set(utcTime);
recycle.setTimezone(tz);
```

I assume in the intention to achieve the same effect, but that does something different, which is the source of the bug being corrected here. Indeed, e.g. if `utcTime` is the timestamp (milliseconds since epoch) for e.g. 2023-08-28 00:00:00 UTC, and TZ is UTC-2 then recycle will contain THE SAME timestamp (the same "absolute" point in time), but expressed in timezone `tz`, that is 2023-08-27 22:00:00 UTC-2. That is not what this function should do. It should return 2023-08-28 00:00:00 UTC-2.

If you prefer, instead of creating `target` fresh each time, we can take __two__ arguments of `Time`, let's call them `recycle0` and `recycle1` and treat them as "object we can freely trash" as we now treat `recycle`

Reply:
It wasn't working before, not with all combinations of timezone and "use home time zone" on/off. See bugs #1173 #1216 #1164 #1208

Reply:
The comment at the top says:

> This (...) will adjust the time to be on a midnight boundary.

The old version clearly _does not adjust the time at all_, and specifically not _to be on a midnight boundary_. My change makes it do what the comment says it does. It adjusts the time to be on a midnight boundary.

One can discuss whether the comment or the code fits the "Google programmer's" intention better, and maybe you believe the code better expresses the intent of the programmer than the comment. However, my understanding of the general situation with all day events is:

- The date must be _stored_ as a UTC date on a midnight boundary.
- The UI code expects treats stuff in in system (or home when that option is enabled) timezone.
- So the stored timestamp must be converted to system/home timezone for the UI code, and the timestamp (date) that was changed in the UI must be converted to UTC in order to
- The above conversions are not "convert the date and time so that this is the same absolute point in time, expressed in a different timezone" but "take the same day-in-month number, the same month number and the same year number but slap system/home TZ, or UTC TZ respectively, on it"
- E.g. when the system/home timezone is UTC+2, the stored begin date of the all-day event is the number of milliseconds between the epoch and "2023-08-28 00:00:00 UTC+0", then the UI timestamp must be "2023-08-28 00:00:00 UTC+2".
- The methods `convertAlldayUtcToLocal` and `convertAlldayLocalToUTC` are supposed to do the conversions described above + force the time at midnight boundary (as per the comment) as a data sanitisation measure.


Generally, I think you are reading the documentation for `android.text.format.Time`, but the "Time" objects here are `com.android.calendarcommon2.Time`

Looking at the implementation of `switchTimezone` in that class on https://android.googlesource.com/platform/frameworks/opt/calendar/+/master/src/com/android/calendarcommon2/Time.java#210 shows it does not behave as your understanding of the documentation of `android.text.format.Time.switchTimeZone()`; here is the implementation:

```
public void switchTimezone(String timezone) {
        long msBefore = mCalendar.getTimeInMillis();
        mCalendar.setTimeZone(TimeZone.getTimeZone(timezone));
        mCalendar.setTimeInMillis(msBefore);
        mDstChangedByField = -1;
        readFieldsFromCalendar();
```

`mCalendar` is a `java.util.GregorianCalendar`; the call `mCalendar.setTimeInMillis(msBefore);` explicitly sets the time to the same absolute time (the same milliseconds since the epoch in UTC), see documentation of `java.util.GregorianCalendar` on https://developer.android.com/reference/java/util/GregorianCalendar, the method `setTimeinMillis` is inherited from `java.util.Calendar`, and the argument `long millis` is documented as:

millis&emsp;long: the new time in UTC milliseconds from the epoch.

So `switchTimezone` clearly does not do what is my understanding of the conversion that should happen there.